arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -407,35 +412,55 @@ public void setBatchSize(int batchSize) {
   }
 
   private static final class PositionVectorReader extends VectorizedArrowReader {
+    private final Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
+    private final BufferAllocator bufferAllocator = ArrowAllocation.rootAllocator();
+    private final boolean setArrowValidityVector;
     private long rowStart;
+    private int batchSize;
+    private FieldVector vec;
     private NullabilityHolder nulls;
 
+    PositionVectorReader(boolean setArrowValidityVector) {
+      this.setArrowValidityVector = setArrowValidityVector;
+    }
+
     @Override
     public VectorHolder read(VectorHolder reuse, int numValsToRead) {
-      Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
-      FieldVector vec = arrowField.createVector(ArrowAllocation.rootAllocator());
-
-      if (reuse != null) {
-        vec.setValueCount(0);
-        nulls.reset();
+      if (reuse == null) {
+        this.vec = newVector();
+        this.nulls = newNullabilityHolder();
       } else {
-        ((BigIntVector) vec).allocateNew(numValsToRead);
-        for (int i = 0; i < numValsToRead; i += 1) {
-          vec.getDataBuffer().setLong(i * Long.BYTES, rowStart + i);
-        }
-        for (int i = 0; i < numValsToRead; i += 1) {
-          BitVectorHelper.setBit(vec.getValidityBuffer(), i);
+        vec.setValueCount(0);
```

Comment:
do we need to call `nulls.reset();` here when reusing? Also do we need to call `nulls.setNotNulls(0, numValsToRead)` for both cases?

Reply:
I had that initially but then we [discussed](https://github.com/apache/iceberg/pull/3533#discussion_r747722269) that the values are never null so calling `setNotNulls` for each batch seemed redundant. That's why I moved that to the method that inits the nullability holder and simply reuse the same nullability holder. What do you think, @nastra?

Reply:
There is no guarantee that the instance variable `vec` is the same one that backs the reused `VectorHolder`. The method contract is that this will fill the reused vector with data, not that it will reuse a vector if the holder is non-null. I think that the `else` case here should set the _local_ variable `vec` to `reuse.vector()`. Technically, we should probably use the same nullability holder as well, but I think that it is okay to return a new `VectorHolder` with the constant nullability holder instead.

Reply:
While I agree with how the reuse logic should behave theoretically, it is not we have in `VectorizedArrowReader`. I looked at that part as an example of how it should behave. I am a bit worried about unforeseen consequences of changing just this place.

Reply:
In `VectorizedArrowReader`, we also create a new `VectorHolder` even if the reuse is enabled.

Reply:
A new `VectorHolder` is fine and shouldn't hurt anything. We need to eventually fix the other places as well, but I don't see a reason why it would hurt to have this place implement the contract properly.